hazelcast/src/test/java/com/hazelcast/client/map/ClientMapIterableTest.java

```diff
@@ -0,0 +1,134 @@
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import java.util.Iterator;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static junit.framework.Assert.assertFalse;
+import static junit.framework.Assert.assertTrue;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ClientMapIterableTest extends HazelcastTestSupport {
+
+    @Parameter
+    public boolean prefetchValues;
+
+    @Parameters(name = "prefetchValues:{0}")
+    public static Iterable<Object[]> parameters() {
+        return asList(new Object[] {Boolean.TRUE}, new Object[] {Boolean.FALSE});
+    }
+
+    protected TestHazelcastFactory factory;
+    protected HazelcastInstance server;
+    protected HazelcastInstance client;
+
+    @Before
+    public void setup() {
+        Config config = getConfig();
+        ClientConfig clientConfig = getClientConfig();
+
+        factory = new TestHazelcastFactory();
+        server = factory.newHazelcastInstance(config);
+        client = factory.newHazelcastClient(clientConfig);
+    }
+
+    @After
+    public void teardown() {
+        factory.terminateAll();
+    }
+
+    @Test
+    public void test_HasNext_Returns_False_On_EmptyMap() throws Exception {
+        ClientMapProxy<Integer, Integer> map = getMapProxy();
+
+        Iterator<Map.Entry<Integer, Integer>> iterator = map.iterable(10, prefetchValues).iterator();
+        assertFalse(iterator.hasNext());
+    }
+
+    @Test
+    public void test_HasNext_Returns_True_On_NonEmptyMap() throws Exception {
+        ClientMapProxy<String, String> map = getMapProxy();
+
+        String key = generateKeyForPartition(server, 1);
+        String value = randomString();
+        map.put(key, value);
+
+        Iterator<Map.Entry<String, String>> iterator = map.iterable(10, prefetchValues).iterator();
+        assertTrue(iterator.hasNext());
+    }
+
+    @Test
+    public void test_Next_Returns_Value_On_NonEmptyMap() throws Exception {
+        ClientMapProxy<String, String> map = getMapProxy();
+
+        String key = generateKeyForPartition(server, 1);
+        String value = randomString();
+        map.put(key, value);
+
+        Iterator<Map.Entry<String, String>> iterator = map.iterable(10, prefetchValues).iterator();
+        Map.Entry entry = iterator.next();
+        assertEquals(value, entry.getValue());
+    }
+
+    @Test
+    public void test_Next_Returns_Values_When_FetchSizeExceeds_On_NonEmptyMap() throws Exception {
+        ClientMapProxy<String, String> map = getMapProxy();
+        String value = randomString();
+        int count = 1000;
+        for (int i = 0; i < count; i++) {
+            String key = generateKeyForPartition(server, 42);
```

Comment:
To improve the test a little bit, can we also make the key randomly distributed over all partitions? I am not sure why we need keys to be owned by a specific partition here